Adding module that manages graph database clusters in AWS Neptune

slapula opened the pull request:
##### SUMMARY
AWS just released a new managed graph database service called AWS Neptune.  I had some time last night to play around with the API and it's similar to RDS and Redshift.  This PR adds a module that manages graph database clusters in Neptune.  Like my Secrets Manager module (#40093), this one also requires `botocore>=1.10.0`.  This will likely be the first of several modules I will be writing for Neptune.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`aws_neptune_cluster`

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = None
  configured module search path = [u'/Users/asmith/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/asmith/.pyenv/versions/2.7.13/lib/python2.7/site-packages/ansible
  executable location = /Users/asmith/.pyenv/versions/2.7.13/bin/ansible
  python version = 2.7.13 (default, Sep  8 2017, 15:16:38) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

slapula committed:
Adding module that manages graph database clusters in AWS Neptune

slapula committed:
Adding module that manages graph database clusters in AWS Neptune

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+    return False
+
+
+def cluster_ready(client, module, params):
```
This would be a good candidate for a custom waiter. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/waiters.py

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+            cluster_ready(client, module, params)
+            result['db_cluster_arn'] = response['DBCluster']['DBClusterArn']
+            result['changed'] = True
+            return result
```
You can omit this return since it will be returned on line 238.

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+            params['FinalDBSnapshotIdentifier'] = module.params.get('final_snapshot_id')
+        response = client.delete_db_cluster(**params)
+        result['changed'] = True
+        return result
```
This is being returned below too.

slapula committed:
fixing try/expect

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+        result['current_config'] = response['DBClusters'][0]
+        result['db_cluster_arn'] = response['DBClusters'][0]['DBClusterArn']
+        return True
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError):
```
Returning False for any of these exceptions can mask important errors. Can you just catch the one specific to the DB cluster not existing to return False and other exceptions call fail_json_aws? Here's an example from another module for how to handle a specific exception code differently:
https://github.com/ansible/ansible/blob/42368ce3101309a51c0f29320d296328c28df918/lib/ansible/modules/cloud/amazon/rds_instance_facts.py#L366-L369

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+
+    if desired_state == 'present':
+        if not cluster_status:
+            create_cluster(client, module, params, result)
```
Rather than initialize `result` above and then passing it around and modifying the contents it would be more clear to return a result dict created inside create_cluster (and update_cluster and delete_cluster) `return {'db_cluster_arn': response['DBCluster']['DBClusterArn'], 'changed': True}` and `result = create_cluster(client, module, params)` (because right now the return statements do nothing), etc

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, AWSRetry
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict, boto3_tag_list_to_ansible_dict
```
This and the line above are unused imports

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+
+DOCUMENTATION = r'''
+---
+module: neptune_cluster
```
Maybe aws_neptune_cluster

slapula committed:
updates based on feedback

slapula committed:
adding custom neptune waiters

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,342 @@
+description:
+    - Create, modify, and destroy graph database clusters on AWS Neptune
+version_added: "2.7"
+requirements: [ 'botocore>=1.10.0', 'boto3' ]
```
I think the minimum botocore is 1.10.30

s-hertel commented on lib/ansible/modules/cloud/amazon/neptune_cluster.py:
```diff
@@ -0,0 +1,342 @@
+        response = client.describe_db_clusters(
+            DBClusterIdentifier=params['DBClusterIdentifier']
+        )
+    except client.exceptions.from_code('DBClusterNotFound'):
```
Sorry, I was mistaken about this, it masks unwanted things. Catching ClientError exceptions and as e and then checking e.response['Error']['Code'] is necessary.

s-hertel commented on test/runner/requirements/constraints.txt:
```diff
@@ -17,3 +17,4 @@ ntlm-auth >= 1.0.6 # message encryption support
 requests-ntlm >= 1.1.0 # message encryption support
 requests-credssp >= 0.1.0 # message encryption support
 voluptuous >= 0.11.0 # Schema recursion via Self
+botocore >= 1.10.0 # adds support for the following AWS services: secretsmanager, fms, acm-pca, neptune
```
Not necessary to add this here. This will mean people not running AWS tests will also install botocore. There's an AWS requirements file which lists boto3, which also installs botocore.

slapula replied:
Understood! That line has been removed.

s-hertel committed:
make requested changes

s-hertel commented:
Is there a reason not to rename it? It would be more consistent with other new AWS modules. Besides that this looks good to me. I expect that it will fail CI right now due to lack of permissions but I'll work on getting those added this week or next.

s-hertel commented:
Thanks for adding the waiters, btw. The botocore repo may be interested in those as well.

slapula committed:
forgot the renamed files

slapula committed:
removing old module

slapula commented:
@s-hertel No good reason, I tend to prefer shorter module names so that's what I went with when I started working on the Neptune stuff. I realize that's not the direction Ansible wants to go with so I've moved this module and the other ones to `aws_*`.  Anyway, this should be ready for another look.  I went ahead and added check_mode support too since I forgot to include it here too.

s-hertel commented:
Hm. Looks like I'm mistaken about test/runner/requirements/integration.cloud.aws.txt... @mattclay should AWS test requirements go in `test/runner/requirements/constraints.txt`? I based my assumption off the azure requirements in `test/runner/requirements/integration.cloud.azure.txt`

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_neptune_cluster.py:
```diff
@@ -0,0 +1,356 @@
+
+
+def cluster_exists(client, module, params):
+    if module.check_mode and module.params.get('state') == 'absent':
```
Oops, missed something. I think this should still return the describe_db_clusters call so changed is accurately reflected by whether or not the cluster exists. No changes are being made by that call so it's fine and you're checking for module.check_mode in delete_cluster().

slapula replied:
Ok, makes sense to me!  Updated.

slapula committed:
removing check_mode if from cluster_exists

mattclay commented on test/runner/requirements/integration.cloud.aws.txt:
```diff
@@ -1,2 +1,3 @@
 boto
 boto3
+botocore>=1.10.30
```
Requirements such as `botocore` should be listed in this file, but version constraints such as `botocore >= 1.10.30` should be in `constraints.txt` instead. Be sure to include a comment explaining the reason for the constraint.

slapula committed:
moving constraint back original file

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_neptune_cluster.py:
```diff
@@ -0,0 +1,354 @@
+            'db_engine_version': dict(type='str', choices=['1.0.1']),
+            'port': dict(type='int', default=8182),
+            'master_username': dict(type='str'),
+            'master_password': dict(type='str'),
```
You should set no_log=True for this

s-hertel reviewed (commented):
Looks great, thanks for the time you spent on this!

s-hertel committed:
Use no_log for password

ryansb commented on test/runner/requirements/constraints.txt:
```diff
@@ -17,3 +17,4 @@ ntlm-auth >= 1.0.6 # message encryption support
 requests-ntlm >= 1.1.0 # message encryption support
 requests-credssp >= 0.1.0 # message encryption support
 voluptuous >= 0.11.0 # Schema recursion via Self
+botocore >= 1.10.30 # adds support for the following AWS services: secretsmanager, fms, acm-pca, neptune
```
I think this affects all integration tests. Instead, I think it belongs in `integration.cloud.aws.txt`. @mattclay can you confirm?

mattclay replied:
This is the correct place to put version constraints. It affects all tests, but will not require the package to be installed for all tests. It will, however, require the version constraint be met whenever the package is installed.

Package requirements (only the package name) should be in the appropriate requirements file, such as `integration.cloud.aws.txt`. However, package constraints (version number requirements) should be placed in `constraints.txt`. This avoids having conflicting version requirements across tests while still allowing different tests to require different packages.

Akasurde commented:
Hi @slapula, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.aws<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
